[mergeTree]: fail fast if max_rows_to_read limit exceeded on parts scan

hagen1778 opened the pull request:
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Performance Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fail fast if `max_rows_to_read` limit is exceeded on parts scan. The motivation behind this change is to skip ranges scan for all selected parts if it is clear that `max_rows_to_read` is already exceeded. The change is quite noticeable for queries over big number of parts.


Detailed description / Documentation draft:

We heavily rely on `max_rows_to_read` limit in our systems and use limit exceptions from CH to decide what to do next: try different table, host, whatever. Row limit is especially important because these limits are evaluated by ClickHouse before any data is read from disk (using only the data from the indexes in-memory). But sometimes we observe significant delays for queries that expected to fail immediately (selecting big time range or too many PK entries), which is unexpected. We also noticed that "delay" before we get exception depends on number of parts query selects. Further investigation discovered that delay caused by fetching ranges for selected parts even if already fetched data is enough to throw an exception.

The change does exactly this - checks the `max_rows_to_read` during the marks and ranges fetch and fails immediately after number of rows exceeds the limit without waiting for all parts to complete.

Commit:
[mergeTree]: fail fast if max_rows_to_read limit exceeded on parts scan

The motivation behind this change is to skip ranges scan for all selected parts
if it is clear that `max_rows_to_read` is already exceeded. The change is quite
noticeable for queries over big number of parts.

alexey-milovidov commented on src/Storages/MergeTree/MergeTreeDataSelectExecutor.cpp:
```diff
@@ -573,6 +574,9 @@ Pipes MergeTreeDataSelectExecutor::readFromParts(
 
     /// Let's find what range to read from each part.
     {
+        std::mutex total_rows_mutex;
```
We can do all we need with std::atomic and single add_and_fetch operation.

alexey-milovidov commented:
The optimization LGTM.

hagen1778 commented:
cc @nvartolomei @kvap

nvartolomei commented on src/Storages/MergeTree/MergeTreeDataSelectExecutor.cpp:
```diff
@@ -599,7 +602,20 @@ Pipes MergeTreeDataSelectExecutor::readFromParts(
+                    if (total_rows_estimate > settings.max_rows_to_read)
+                        throw Exception(
+                            "Limit for rows exceeded, max rows: " + formatReadableQuantity(settings.max_rows_to_read)
+                            + ", current rows: " + formatReadableQuantity(total_rows_estimate),
```
nit: total_row_estimate in this error message might be a bit miss-leading as it can be interpreted by the user as "limit 100 but need to read $total_row_estimate". Is it much more expensive if this is moved after all process_part are executed so that the actual number of estimated rows could be printed in the error (this would be consistent with current limit logic irc)? If it would be expensive I'd just remove this part of the message.

alexey-milovidov replied:
`current rows` -> `estimated rows to read (at least)`
And also would be nice to add the name of the setting:
`Limit for rows (controlled by `max_rows_to_read` setting) exceeded`

kvap commented on src/Storages/MergeTree/MergeTreeDataSelectExecutor.cpp:
```diff
@@ -599,7 +602,20 @@ Pipes MergeTreeDataSelectExecutor::readFromParts(
+                if (settings.read_overflow_mode == OverflowMode::THROW && settings.max_rows_to_read)
+                {
+                    /// Fail fast if estimated number of rows to read exceeds the limit
+                    size_t total_rows_estimate = total_rows.fetch_add(ranges.getRowsCount());
```
`fetch_add` returns the value that was there before the addition. You can make it fail a little bit earlier if you compare the estimate after incrementing.

```cpp
auto part_rows_estimate = ranges.getRowsCount();
size_t prev_total_rows_estimate = total_rows.fetch_add(parts_rows_estimate);
if (prev_total_rows_estimate + part_rows_estimate > settings.max_rows_to_read) { ... }
```

Commit:
[mergeTree]: make exception message more clear

hagen1778 commented:
@alexey-milovidov this change breaks test case [01064](https://github.com/ClickHouse/ClickHouse/blob/master/tests/queries/0_stateless/01064_incremental_streaming_from_2_src_with_feedback.sql#L90) which relies on exact number of read rows, not estimated. Would it be better if we use a separate setting for this behaviour like `max_rows_to_read_estimated` or `estimated_max_rows_to_read`?

alexey-milovidov commented:
Maybe adjusting this test is Ok.
But why the number of estimated rows is large?

Shouldn't it be filtered by index:
```
FROM target_table
        WHERE id IN (SELECT id FROM logins)
```
?

hagen1778 commented:
> But why the number of estimated rows is large?

I understand it as following:
1. Table `target_table` [created](https://github.com/ClickHouse/ClickHouse/blob/master/tests/queries/0_stateless/01064_incremental_streaming_from_2_src_with_feedback.sql#L14) and populated with 50k rows - forms 1 part on disk
2. Merges are [disabled](https://github.com/ClickHouse/ClickHouse/blob/master/tests/queries/0_stateless/01064_incremental_streaming_from_2_src_with_feedback.sql#L79) then - every new insert will form a new part on disk
3. [Insert](https://github.com/ClickHouse/ClickHouse/blob/master/tests/queries/0_stateless/01064_incremental_streaming_from_2_src_with_feedback.sql#L84) into `logins` forms 2nd part
4. [Insert](https://github.com/ClickHouse/ClickHouse/blob/master/tests/queries/0_stateless/01064_incremental_streaming_from_2_src_with_feedback.sql#L85) into `checkouts` forms 3d part

Hence, attempt to insert into `logins` after limit was set will have to read at least 3 parts and 3 marks with 8192 granularity. According to this, test may be explained in following way:
```

-- three parts to read, means three marks with 8192 granularity at least
set max_rows_to_read = 24576;
INSERT INTO logins    SELECT number as id, '2000-01-01 11:00:00' from numbers(1000);

-- four parts to read, means four marks with 8192 granularity at least
set max_rows_to_read = 32768;
INSERT INTO checkouts SELECT number as id, '2000-01-01 11:10:00' from numbers(1000);

-- five parts to read, means five marks with 8192 granularity at least
set max_rows_to_read = 40960;
INSERT INTO logins    SELECT number+2 as id, '2001-01-01 11:10:01' from numbers(1);

-- six parts to read, means six marks with 8192 granularity at least
set max_rows_to_read = 49152;
INSERT INTO checkouts SELECT number+2 as id, '2001-01-01 11:10:02' from numbers(1);
```
Does that make sense?

alexey-milovidov commented:
Yes,
let's adjust the test in this PR.

filimonov commented:
The main idea of that test is that MV should not 'see'/'read' real table but only inserted block (even if lef-most appaets in MV several times).

filimonov commented:
Meybe optimization is fired before replacement of "real" table to a "inserted blocks only" happens (similar to https://github.com/ClickHouse/ClickHouse/pull/10603/ ) ?

hagen1778 commented:
> May be optimization is fired before replacement of "real" table to an "inserted blocks only" happens (similar to #10603 ) ?

Could be, I'm not sure. Probably statement in `LEFT JOIN` scans "real" data.

```
...
 LEFT JOIN (
    SELECT
        id,
        maxMerge(latest_checkout_time) as current_latest_checkout_time
    FROM target_table
    WHERE id IN (SELECT id FROM logins)
    GROUP BY id
    ) USING (id)
```

That matches with tracing logs on insert:
```
[C02WQ331HTDD] 2020.08.17 15:19:58.665347 [ 5683920 ] {5e888d39-4b99-4ac9-8c6a-cd8a9c49b9e3} <Debug> executeQuery: (from [::1]:65420) INSERT INTO logins SELECT number as id, '2000-01-01 11:00:00' from numbers(1000);
...
[C02WQ331HTDD] 2020.08.17 15:19:58.745386 [ 5684270 ] {5e888d39-4b99-4ac9-8c6a-cd8a9c49b9e3} <Debug> HashJoin: Right sample block: id UInt64 UInt64(size = 0), current_latest_checkout_time DateTime('UTC') UInt32(size = 0)
[C02WQ331HTDD] 2020.08.17 15:19:58.750453 [ 5684270 ] {5e888d39-4b99-4ac9-8c6a-cd8a9c49b9e3} <Trace> InterpreterSelectQuery: FetchColumns -> Complete
[C02WQ331HTDD] 2020.08.17 15:19:58.757933 [ 5684270 ] {5e888d39-4b99-4ac9-8c6a-cd8a9c49b9e3} <Trace> CreatingSetsBlockInputStream: Creating join.
[C02WQ331HTDD] 2020.08.17 15:19:58.760121 [ 5684270 ] {5e888d39-4b99-4ac9-8c6a-cd8a9c49b9e3} <Debug> default.target_table (SelectExecutor): Key condition: (column 0 in 1000-element set)
[C02WQ331HTDD] 2020.08.17 15:19:58.763775 [ 5684271 ] {5e888d39-4b99-4ac9-8c6a-cd8a9c49b9e3} <Trace> default.target_table (SelectExecutor): Used generic exclusion search over index for part all_1_1_0 with 8 steps
[C02WQ331HTDD] 2020.08.17 15:19:58.763961 [ 5684273 ] {5e888d39-4b99-4ac9-8c6a-cd8a9c49b9e3} <Trace> default.target_table (SelectExecutor): Used generic exclusion search over index for part all_2_2_0 with 8 steps
[C02WQ331HTDD] 2020.08.17 15:19:58.765434 [ 5684270 ] {5e888d39-4b99-4ac9-8c6a-cd8a9c49b9e3} <Debug> default.target_table (SelectExecutor): Selected 2 parts by date, 2 parts by key, 2 marks by primary key, 2 marks to read from 2 ranges
[C02WQ331HTDD] 2020.08.17 15:19:58.767000 [ 5684270 ] {5e888d39-4b99-4ac9-8c6a-cd8a9c49b9e3} <Trace> default.target_table (SelectExecutor): Reading approx. 16384 rows with 2 streams

```

filimonov commented:
> Do you think there is no sense in adjusting the test case, @filimonov ?

Actually i like that idea:

> use a separate setting for this behaviour like `max_rows_to_read_estimated` or `estimated_max_rows_to_read`?

Motivation -

1) currently it can fail after sending some data (we never know how users use the settings and don't want to introduce incompatinilities unless it's necessary):

For example:
```
set max_block_size=1, max_rows_to_read=50;
select * from numbers_mt(100);
```

Also, the current limit is based on the actual rows, not predicted (not sure if the prediction is always accurate - with all types of filtering, joins, unions, etc). 

BTW: that reminds me of the similar issue with max_execution_time / vs estimated_max_execution_time introduced in 20.4.

KochetovNicolai commented:
I think that:
1. We don't need a separates setting for `select` like `max_rows_to_read_estimated` or `estimated_max_rows_to_read` because it would be extremely difficult to explain what it does and why it is needed. `max_rows_to_read` fits perfect here.
2. We shouldn't fail because of `max_rows_to_read` while inserting data. It is not obvious. `max_rows_to_read` should be ignored. It is possible to introduce a new setting like `max_rows_to_read_at_insert_into_materialized_view` or something, which may be or may not be equal to `max_rows_to_read`. 
3. Test should be fixed. We may read the number of rows form `system.query_log` or specify a new setting.

Also I don't understand why test 01064 did not fail before. My assumption that condition `WHERE id IN (SELECT id FROM logins)` was moved to `prewhere`, and `max_rows_to_read` seem to be calculated after `prewhere` (before changes in this pr) (which, I suppose, is not quite correct). If it is true, we need to make this test to hits pk index better.

Commit:
[tests]: adjust staeless/01064 test `max_rows_to_read` limits

With default index_granularity=8096 CH always reads at least that number
of rows per selected part, but limit is checked afterwards. The optimization
that interrupts execution of queries based on approx number of rows to read
breaks the test. This means that test case is potetntially contains incorrect
limits.
